File: storybook/lua/animate.lua
--[[
  animate.lua -- Lua-side transition and animation engine

  Follows the TextEditor "Lua-owns-interaction" pattern: JS declares targets
  via style props, Lua interpolates autonomously. Zero per-frame bridge traffic
  for active animations.

  Architecture:
    - JS sends `transition` config as part of the style (e.g., transition = {
        backgroundColor = { duration = 300, easing = "easeInOut" } })
    - tree.lua detects style changes on UPDATE and calls Animate.processStyleUpdate()
    - For transitioned properties, Animate stores {from, to, startTime, ...} and
      writes the interpolated value to node.style each frame
    - Painter reads node.style as normal -- no painter changes needed
    - Visual-only properties (opacity, colors, transform) skip re-layout
    - Layout-affecting properties (width, height, padding, margin) mark tree dirty

  Supports:
    - CSS transition: per-property duration, easing, delay
    - Easing: linear, easeIn, easeOut, easeInOut, bezier, bounce, elastic
    - Color interpolation: hex strings lerped in RGBA space
    - Transform interpolation: per-component lerp (translate, rotate, scale)
    - Percentage string interpolation: "50%" -> "75%"
    - Keyframe animations: placeholder for Phase B
]]

local Animate = {}

-- ============================================================================
-- State
-- ============================================================================

local activeNodes = {}   -- nodeId -> node (nodes with running transitions/animations)
local treeRef = nil      -- reference to tree module for markDirty()

-- ============================================================================
-- Init
-- ============================================================================

--- Initialize the animation module.
--- @param config table|nil  { tree = TreeModule }
function Animate.init(config)
  config = config or {}
  treeRef = config.tree
  activeNodes = {}
end

-- ============================================================================
-- Easing functions (ported from packages/shared/src/animation.ts)
-- ============================================================================

local easing = {}

easing.linear = function(t) return t end

easing.easeIn = function(t) return t * t end

easing.easeOut = function(t) return t * (2 - t) end

easing.easeInOut = function(t)
  if t < 0.5 then return 2 * t * t end
  return -1 + (4 - 2 * t) * t
end

-- Cubic bezier helpers
local function cubicBezier(t, p1, p2)
  local mt = 1 - t
  return 3 * mt * mt * t * p1 + 3 * mt * t * t * p2 + t * t * t
end

local function cubicBezierDerivative(t, p1, p2)
  local mt = 1 - t
  return 3 * mt * mt * p1 + 6 * mt * t * (p2 - p1) + 3 * t * t * (1 - p2)
end

--- Create a cubic bezier easing function.
--- @param x1 number Control point 1 X
--- @param y1 number Control point 1 Y
--- @param x2 number Control point 2 X
--- @param y2 number Control point 2 Y
--- @return function Easing function
function easing.bezier(x1, y1, x2, y2)
  return function(t)
    if t <= 0 then return 0 end
    if t >= 1 then return 1 end
    -- Newton-Raphson: solve for u where bezierX(u) = t
    local u = t
    for _ = 1, 8 do
      local xEst = cubicBezier(u, x1, x2) - t
      if math.abs(xEst) < 1e-6 then break end
      local dx = cubicBezierDerivative(u, x1, x2)
      if math.abs(dx) < 1e-6 then break end
      u = u - xEst / dx
    end
    u = math.max(0, math.min(1, u))
    return cubicBezier(u, y1, y2)
  end
end

easing.bounce = function(t)
  if t < 1 / 2.75 then
    return 7.5625 * t * t
  elseif t < 2 / 2.75 then
    local t2 = t - 1.5 / 2.75
    return 7.5625 * t2 * t2 + 0.75
  elseif t < 2.5 / 2.75 then
    local t2 = t - 2.25 / 2.75
    return 7.5625 * t2 * t2 + 0.9375
  else
    local t2 = t - 2.625 / 2.75
    return 7.5625 * t2 * t2 + 0.984375
  end
end

--- Create an elastic easing function.
--- @param bounciness number|nil Bounciness factor (default 1)
--- @return function Easing function
function easing.elastic(bounciness)
  bounciness = bounciness or 1
  local p = 0.3 / math.max(bounciness, 0.001)
  return function(t)
    if t <= 0 then return 0 end
    if t >= 1 then return 1 end
    return math.pow(2, -10 * t) * math.sin(((t - p / 4) * (2 * math.pi)) / p) + 1
  end
end

--- Resolve an easing specifier to a function.
--- Accepts: string name, function, or table { type="bezier", [1]=x1, [2]=y1, ... }
--- @param e any Easing specifier
--- @return function
local function resolveEasing(e)
  if type(e) == "function" then return e end
  if type(e) == "string" then
    -- Handle "elastic" with default bounciness
    if e == "elastic" then return easing.elastic(1) end
    return easing[e] or easing.easeInOut
  end
  if type(e) == "table" then
    -- Bezier: { type = "bezier", x1, y1, x2, y2 } or just { x1, y1, x2, y2 }
    if e.type == "bezier" or (#e == 4 and type(e[1]) == "number") then
      return easing.bezier(e[1] or e.x1 or 0, e[2] or e.y1 or 0,
                           e[3] or e.x2 or 1, e[4] or e.y2 or 1)
    end
    -- Elastic with custom bounciness: { type = "elastic", bounciness = 2 }
    if e.type == "elastic" then
      return easing.elastic(e.bounciness or 1)
    end
  end
  return easing.easeInOut
end

-- ============================================================================
-- Color parsing and interpolation
-- ============================================================================

--- Parse a color value to RGBA components (0-1 range).
--- Accepts hex strings ("#rrggbb", "#rrggbbaa"), "transparent", or {r,g,b,a} tables.
--- @param c any Color value
--- @return number|nil r, number|nil g, number|nil b, number|nil a
local function parseColor(c)
  if type(c) == "table" then
    return c[1] or 0, c[2] or 0, c[3] or 0, c[4] or 1
  end
  if type(c) ~= "string" then return nil end
  if c == "transparent" then return 0, 0, 0, 0 end
  local r, g, b, a = c:match("#(%x%x)(%x%x)(%x%x)(%x?%x?)")
  if r then
    local alpha = 1
    if a and a ~= "" then alpha = tonumber(a, 16) / 255 end
    return tonumber(r, 16) / 255, tonumber(g, 16) / 255, tonumber(b, 16) / 255, alpha
  end
  return nil
end

--- Convert RGBA components (0-1) to a hex color string.
--- @param r number Red 0-1
--- @param g number Green 0-1
--- @param b number Blue 0-1
--- @param a number Alpha 0-1
--- @return string Hex color string
local function colorToHex(r, g, b, a)
  local ri = math.floor(r * 255 + 0.5)
  local gi = math.floor(g * 255 + 0.5)
  local bi = math.floor(b * 255 + 0.5)
  if a ~= nil and a < 0.999 then
    local ai = math.floor(a * 255 + 0.5)
    return string.format("#%02x%02x%02x%02x", ri, gi, bi, ai)
  end
  return string.format("#%02x%02x%02x", ri, gi, bi)
end

-- ============================================================================
-- Value interpolation
-- ============================================================================

--- Linearly interpolate between two numbers.
local function lerp(a, b, t)
  return a + (b - a) * t
end

--- Properties that require color interpolation.
local colorProps = {
  backgroundColor = true,
  borderColor = true,
  borderTopColor = true,
  borderRightColor = true,
  borderBottomColor = true,
  borderLeftColor = true,
  shadowColor = true,
  color = true,
}

--- Properties that are visual-only and do NOT affect layout.
--- Transitions on these skip tree.markDirty() for performance.
local visualOnlyProps = {
  opacity = true,
  backgroundColor = true,
  borderColor = true,
  borderTopColor = true,
  borderRightColor = true,
  borderBottomColor = true,
  borderLeftColor = true,
  shadowColor = true,
  shadowOffsetX = true,
  shadowOffsetY = true,
  shadowBlur = true,
  color = true,
  transform = true,
  zIndex = true,
  backgroundGradient = true,
}

--- Interpolate a transform object component by component.
--- @param from table|nil Source transform
--- @param to table|nil Target transform
--- @param t number Progress 0-1
--- @return table Interpolated transform
local function lerpTransform(from, to, t)
  if not from and not to then return nil end
  from = from or {}
  to = to or {}
  local result = {}

  local keys = { "translateX", "translateY", "rotate", "scaleX", "scaleY", "originX", "originY" }
  for _, key in ipairs(keys) do
    local fv = from[key]
    local tv = to[key]
    if fv ~= nil or tv ~= nil then
      -- Defaults: scale=1, origin=0.5, everything else=0
      local defaultVal = 0
      if key == "scaleX" or key == "scaleY" then defaultVal = 1
      elseif key == "originX" or key == "originY" then defaultVal = 0.5
      end
      fv = fv or defaultVal
      tv = tv or defaultVal
      result[key] = lerp(fv, tv, t)
    end
  end

  return result
end

--- Interpolate a single style value based on its property name.
--- Handles numbers, colors, transforms, and percentage strings.
--- @param from any Source value
--- @param to any Target value
--- @param t number Progress 0-1
--- @param propName string The style property name
--- @return any Interpolated value
local function interpolateValue(from, to, t, propName)
  -- Color interpolation
  if colorProps[propName] then
    local r1, g1, b1, a1 = parseColor(from)
    local r2, g2, b2, a2 = parseColor(to)
    if r1 and r2 then
      return colorToHex(lerp(r1, r2, t), lerp(g1, g2, t), lerp(b1, b2, t), lerp(a1, a2, t))
    end
    -- Cannot interpolate, snap at midpoint
    if t < 0.5 then return from else return to end
  end

  -- Transform interpolation
  if propName == "transform" then
    return lerpTransform(from, to, t)
  end

  -- Numeric interpolation
  if type(from) == "number" and type(to) == "number" then
    return lerp(from, to, t)
  end

  -- Percentage string interpolation ("50%" -> "75%")
  if type(from) == "string" and type(to) == "string" then
    local fNum = tonumber(from:match("^([%d%.]+)%%$"))
    local tNum = tonumber(to:match("^([%d%.]+)%%$"))
    if fNum and tNum then
      return string.format("%.1f%%", lerp(fNum, tNum, t))
    end
  end

  -- Cannot interpolate, snap at midpoint
  if t < 0.5 then return from else return to end
end

-- ============================================================================
-- Transition processing
-- ============================================================================

--- Process a style update on a node, setting up transitions for properties
--- that have a transition config.
---
--- Called by tree.lua AFTER the style diff has been applied to node.style.
--- For transitioned properties, this function writes the "from" value back
--- to node.style (so the visual doesn't jump) and sets up the interpolation
--- state. animate.tick() will update node.style each frame.
---
--- @param node table The tree node
--- @param oldValues table Map of propName -> previous value (before this UPDATE)
--- @param newValues table Map of propName -> new value (the style diff from JS)
function Animate.processStyleUpdate(node, oldValues, newValues)
  local transConfig = node.style and node.style.transition
  if not transConfig then return end

  for propName, newValue in pairs(newValues) do
    -- Skip config properties themselves
    if propName ~= "transition" and propName ~= "animation" then
      local oldValue = oldValues[propName]

      -- Only transition if value actually changed and we have a previous value
      if oldValue ~= nil and oldValue ~= newValue then
        -- Check for per-property or "all" config
        local config = transConfig[propName] or transConfig.all
        if config and type(config) == "table" then
          -- Set up transition state
          if not node.transitionState then node.transitionState = {} end

          -- If a transition is already running, start from its current interpolated value
          local existing = node.transitionState[propName]
          local fromValue = oldValue
          if existing and existing.current ~= nil then
            fromValue = existing.current
          end

          node.transitionState[propName] = {
            from = fromValue,
            to = newValue,
            startTime = love.timer.getTime(),
            duration = (config.duration or 300) / 1000,  -- ms -> seconds
            easing = config.easing or "easeInOut",
            delay = (config.delay or 0) / 1000,          -- ms -> seconds
            current = fromValue,
          }

          -- Write back the from value so the visual doesn't jump
          node.style[propName] = fromValue

          -- Register this node as active
          activeNodes[node.id] = node
        end
      end
    end
  end
end

--- Called when a node is removed from the tree.
--- Cleans up any active transitions/animations.
--- @param nodeId number|string The node ID being removed
function Animate.onNodeRemoved(nodeId)
  activeNodes[nodeId] = nil
end

-- ============================================================================
-- Per-frame tick
-- ============================================================================

--- Update all active transitions and animations.
--- Call once per frame from init.lua, AFTER tree.applyCommands and BEFORE layout.
--- @param dt number Delta time in seconds (from love.update)
--- @return boolean True if any animations are still active
function Animate.tick(dt)
  local now = love.timer.getTime()
  local toRemove = {}
  local needsLayout = false

  for nodeId, node in pairs(activeNodes) do
    local allDone = true

    -- ── Process transitions ──
    if node.transitionState then
      for propName, ts in pairs(node.transitionState) do
        local elapsed = now - ts.startTime

        if elapsed < ts.delay then
          -- Still in delay period
          allDone = false
        else
          local activeElapsed = elapsed - ts.delay
          local progress = 1
          if ts.duration > 0 then
            progress = math.min(activeElapsed / ts.duration, 1)
          end

          local easingFn = resolveEasing(ts.easing)
          local easedProgress = easingFn(progress)
          local value = interpolateValue(ts.from, ts.to, easedProgress, propName)

          ts.current = value
          node.style[propName] = value

          if progress >= 1 then
            -- Transition complete: snap to final value and clean up
            node.style[propName] = ts.to
            node.transitionState[propName] = nil
          else
            allDone = false
          end

          -- Layout-affecting properties need tree relayout
          if not visualOnlyProps[propName] then
            needsLayout = true
          end
        end
      end

      -- Clean up empty transitionState table
      if allDone then
        local hasAny = false
        for _ in pairs(node.transitionState) do hasAny = true; break end
        if not hasAny then node.transitionState = nil end
      end
    end

    -- ── Process keyframe animations (Phase B placeholder) ──
    if node.animationState then
      local as = node.animationState
      if as.playState ~= "paused" then
        local elapsed = now - as.startTime

        if elapsed < (as.delay or 0) then
          allDone = false
        else
          local activeElapsed = elapsed - (as.delay or 0)
          local iterationProgress = 0
          if as.duration > 0 then
            iterationProgress = activeElapsed / as.duration
          end

          -- Handle iterations
          local currentIteration = math.floor(iterationProgress)
          local withinIteration = iterationProgress - currentIteration

          -- Check completion
          local iterations = as.iterations or 1
          local isComplete = iterations > 0 and currentIteration >= iterations
          if isComplete then
            withinIteration = 1
            currentIteration = iterations - 1
          end

          -- Apply direction
          local direction = as.direction or "normal"
          local progress = withinIteration
          if direction == "reverse" then
            progress = 1 - progress
          elseif direction == "alternate" then
            if currentIteration % 2 == 1 then progress = 1 - progress end
          elseif direction == "alternate-reverse" then
            if currentIteration % 2 == 0 then progress = 1 - progress end
          end

          -- Apply easing
          local easingFn = resolveEasing(as.easing)
          local easedProgress = easingFn(progress)

          -- Interpolate keyframe values
          local keyframes = as.keyframes
          if keyframes then
            -- Find bounding keyframe stops
            local stops = as.sortedStops
            if not stops then
              stops = {}
              for pct in pairs(keyframes) do
                stops[#stops + 1] = tonumber(pct) or 0
              end
              table.sort(stops)
              as.sortedStops = stops
            end

            local scaledProgress = easedProgress * 100
            local lowerIdx, upperIdx = 1, #stops
            for i = 1, #stops - 1 do
              if scaledProgress >= stops[i] and scaledProgress <= stops[i + 1] then
                lowerIdx = i
                upperIdx = i + 1
                break
              end
            end

            local lowerPct = stops[lowerIdx]
            local upperPct = stops[upperIdx]
            local segmentT = 0
            if upperPct ~= lowerPct then
              segmentT = (scaledProgress - lowerPct) / (upperPct - lowerPct)
            end

            local lowerFrame = keyframes[tostring(lowerPct)] or keyframes[lowerPct]
            local upperFrame = keyframes[tostring(upperPct)] or keyframes[upperPct]

            if lowerFrame and upperFrame then
              -- Collect all animated property names
              local animProps = {}
              for k in pairs(lowerFrame) do animProps[k] = true end
              for k in pairs(upperFrame) do animProps[k] = true end

              -- Interpolate each property
              if not as.currentValues then as.currentValues = {} end
              for propName in pairs(animProps) do
                local fv = lowerFrame[propName]
                local tv = upperFrame[propName]
                if fv ~= nil and tv ~= nil then
                  local value = interpolateValue(fv, tv, segmentT, propName)
                  as.currentValues[propName] = value
                  node.style[propName] = value
                  if not visualOnlyProps[propName] then
                    needsLayout = true
                  end
                end
              end
            end
          end

          if isComplete then
            -- Apply fill mode
            local fillMode = as.fillMode or "none"
            if fillMode == "none" then
              -- Restore original style values
              if as.originalValues then
                for propName, val in pairs(as.originalValues) do
                  node.style[propName] = val
                end
              end
              node.animationState = nil
            elseif fillMode == "forwards" or fillMode == "both" then
              -- Keep final values (already applied)
              node.animationState = nil
            else
              node.animationState = nil
            end
          else
            allDone = false
          end
        end
      else
        -- Paused: keep node active but don't progress
        allDone = false
      end
    end

    if allDone then
      toRemove[#toRemove + 1] = nodeId
    end
  end

  -- Remove completed nodes from active set
  for _, nodeId in ipairs(toRemove) do
    activeNodes[nodeId] = nil
  end

  -- Mark tree dirty if layout-affecting properties changed
  if needsLayout and treeRef then
    treeRef.markDirty()
  end

  return next(activeNodes) ~= nil
end

-- ============================================================================
-- Animation setup (for keyframe animations via style.animation prop)
-- ============================================================================

--- Set up a keyframe animation on a node from its style.animation config.
--- Called by tree.lua when the animation prop is first set or changes.
--- @param node table The tree node
--- @param animConfig table The animation config from style.animation
function Animate.setupAnimation(node, animConfig)
  if not animConfig or not animConfig.keyframes then return end

  -- Snapshot original values for fillMode: "none" restore
  local originalValues = {}
  for _, frame in pairs(animConfig.keyframes) do
    if type(frame) == "table" then
      for propName in pairs(frame) do
        if originalValues[propName] == nil then
          originalValues[propName] = node.style[propName]
        end
      end
    end
  end

  node.animationState = {
    keyframes = animConfig.keyframes,
    duration = (animConfig.duration or 300) / 1000,    -- ms -> seconds
    easing = animConfig.easing or "linear",
    iterations = animConfig.iterations or 1,           -- -1 = infinite
    direction = animConfig.direction or "normal",
    fillMode = animConfig.fillMode or "none",
    delay = (animConfig.delay or 0) / 1000,            -- ms -> seconds
    startTime = love.timer.getTime(),
    playState = animConfig.playState or "running",
    originalValues = originalValues,
    currentValues = {},
    sortedStops = nil,                                 -- computed lazily
  }

  activeNodes[node.id] = node
end

-- ============================================================================
-- Utility
-- ============================================================================

--- Check if any animations or transitions are currently active.
--- @return boolean
function Animate.hasActiveAnimations()
  return next(activeNodes) ~= nil
end

--- Clear all active animations and transitions.
--- Call on hot reload or tree reset.
function Animate.clear()
  activeNodes = {}
end

--- Expose easing functions for external use.
Animate.easing = easing

--- Expose interpolation for external use (e.g., hover style merging).
Animate.interpolateValue = interpolateValue

--- Expose color utilities for external use.
Animate.parseColor = parseColor
Animate.colorToHex = colorToHex
Animate.lerp = lerp

return Animate

File: tsz/runtime/framework/animate.zig
//! ReactJIT Animation System — Phases 1-3
//!
//! Easing functions (pure math), timing-based animation slots,
//! and spring physics slots. Compile-time allocated, zero heap.
//!
//! Same pattern as state.zig: fixed slot arrays, created at init,
//! ticked each frame. The compiler emits calls to this module
//! for useTransition() and useSpring() hooks.
//!
//! Reference: love2d/lua/animate.lua

const std = @import("std");

// ════════════════════════════════════════════════════════════════════════
// Phase 1: Easing Functions (pure math, zero state)
// Reference: love2d/lua/animate.lua:54-161
// ════════════════════════════════════════════════════════════════════════

pub fn linear(t: f32) f32 {
    return t;
}

pub fn easeIn(t: f32) f32 {
    return t * t;
}

pub fn easeOut(t: f32) f32 {
    return t * (2.0 - t);
}

pub fn easeInOut(t: f32) f32 {
    if (t < 0.5) return 2.0 * t * t;
    return -1.0 + (4.0 - 2.0 * t) * t;
}

/// Spring-like easing with overshoot (decaying sinusoid).
/// Reference: animate.lua:68-73
pub fn spring(t: f32) f32 {
    if (t <= 0) return 0;
    if (t >= 1) return 1;
    const p: f32 = 0.3;
    return std.math.pow(f32, 2.0, -10.0 * t) * @sin((t - p / 4.0) * (2.0 * std.math.pi) / p) + 1.0;
}

/// Bounce easing (multiple bounces).
/// Reference: animate.lua:110-123
pub fn bounce(t: f32) f32 {
    if (t < 1.0 / 2.75) {
        return 7.5625 * t * t;
    } else if (t < 2.0 / 2.75) {
        const t2 = t - 1.5 / 2.75;
        return 7.5625 * t2 * t2 + 0.75;
    } else if (t < 2.5 / 2.75) {
        const t2 = t - 2.25 / 2.75;
        return 7.5625 * t2 * t2 + 0.9375;
    } else {
        const t2 = t - 2.625 / 2.75;
        return 7.5625 * t2 * t2 + 0.984375;
    }
}

/// Elastic easing with configurable bounciness.
/// Reference: animate.lua:128-136
pub fn elastic(t: f32, bounciness: f32) f32 {
    if (t <= 0) return 0;
    if (t >= 1) return 1;
    const b = @max(bounciness, 0.001);
    const p = 0.3 / b;
    return std.math.pow(f32, 2.0, -10.0 * t) * @sin((t - p / 4.0) * (2.0 * std.math.pi) / p) + 1.0;
}

/// Cubic bezier easing via Newton-Raphson (8 iterations).
/// Reference: animate.lua:76-108
pub fn cubicBezier(x1: f32, y1: f32, x2: f32, y2: f32, t: f32) f32 {
    if (t <= 0) return 0;
    if (t >= 1) return 1;

    // Solve for u where bezierX(u) = t using Newton-Raphson
    var u: f32 = t;
    for (0..8) |_| {
        const x_est = bezierCurve(u, x1, x2) - t;
        if (@abs(x_est) < 1e-6) break;
        const dx = bezierDerivative(u, x1, x2);
        if (@abs(dx) < 1e-6) break;
        u -= x_est / dx;
    }
    u = std.math.clamp(u, 0.0, 1.0);
    return bezierCurve(u, y1, y2);
}

fn bezierCurve(t: f32, p1: f32, p2: f32) f32 {
    const mt = 1.0 - t;
    return 3.0 * mt * mt * t * p1 + 3.0 * mt * t * t * p2 + t * t * t;
}

fn bezierDerivative(t: f32, p1: f32, p2: f32) f32 {
    const mt = 1.0 - t;
    return 3.0 * mt * mt * p1 + 6.0 * mt * t * (p2 - p1) + 3.0 * t * t * (1.0 - p2);
}

// ════════════════════════════════════════════════════════════════════════
// Interpolation helpers
// ════════════════════════════════════════════════════════════════════════

/// Linear interpolation between two floats.
pub fn lerp(a: f32, b: f32, t: f32) f32 {
    return a + (b - a) * t;
}

/// Color type matching layout.zig's Color.
pub const Color = struct {
    r: u8,
    g: u8,
    b: u8,
    a: u8,
};

/// Channel-wise color interpolation.
/// Reference: animate.lua:251-261
pub fn lerpColor(a: Color, b: Color, t: f32) Color {
    return .{
        .r = @intFromFloat(lerp(@floatFromInt(a.r), @floatFromInt(b.r), t)),
        .g = @intFromFloat(lerp(@floatFromInt(a.g), @floatFromInt(b.g), t)),
        .b = @intFromFloat(lerp(@floatFromInt(a.b), @floatFromInt(b.b), t)),
        .a = @intFromFloat(lerp(@floatFromInt(a.a), @floatFromInt(b.a), t)),
    };
}

// ════════════════════════════════════════════════════════════════════════
// Easing function type + resolver
// ════════════════════════════════════════════════════════════════════════

pub const EasingFn = *const fn (f32) f32;

/// Resolve an easing name to a function pointer.
/// Used by the compiler to map string literals like 'easeOut' to function pointers.
pub fn resolveEasing(name: []const u8) EasingFn {
    if (std.mem.eql(u8, name, "linear")) return linear;
    if (std.mem.eql(u8, name, "easeIn")) return easeIn;
    if (std.mem.eql(u8, name, "easeOut")) return easeOut;
    if (std.mem.eql(u8, name, "easeInOut")) return easeInOut;
    if (std.mem.eql(u8, name, "spring")) return spring;
    if (std.mem.eql(u8, name, "bounce")) return bounce;
    return easeInOut; // default
}

// ════════════════════════════════════════════════════════════════════════
// Phase 2: Animation Slots (timing-based)
// Reference: animate.lua:360-368 (state), animate.lua:494-510 (tick)
// ════════════════════════════════════════════════════════════════════════

pub const MAX_ANIM_SLOTS = 64;

pub const AnimSlot = struct {
    from: f32 = 0,
    to: f32 = 0,
    current: f32 = 0,
    progress: f32 = 0, // 0.0 → 1.0
    duration_ms: f32 = 300,
    start_time: f32 = 0, // seconds (from SDL_GetTicks / 1000)
    easing: EasingFn = easeInOut,
    active: bool = false,
    done: bool = true,
};

var anim_slots: [MAX_ANIM_SLOTS]AnimSlot = [_]AnimSlot{.{}} ** MAX_ANIM_SLOTS;
var anim_slot_count: usize = 0;
var _anims_active: bool = false;

/// Create a timing-based animation slot. Returns the slot ID.
/// Called once at init time (like state.createSlot).
pub fn createAnim(duration_ms: f32, easing: EasingFn) usize {
    const id = anim_slot_count;
    std.debug.assert(id < MAX_ANIM_SLOTS);
    anim_slots[id] = .{
        .duration_ms = duration_ms,
        .easing = easing,
    };
    anim_slot_count += 1;
    return id;
}

/// Start (or restart) an animation from one value to another.
pub fn startAnim(id: usize, from: f32, to: f32) void {
    var s = &anim_slots[id];
    s.from = from;
    s.to = to;
    s.current = from;
    s.progress = 0;
    s.start_time = -1; // will be set on first tick
    s.active = true;
    s.done = false;
    _anims_active = true;
}

/// Get the current animated value for a slot.
pub fn getAnimValue(id: usize) f32 {
    return anim_slots[id].current;
}

/// Get the target value for a slot (used to detect target changes).
pub fn getAnimTarget(id: usize) f32 {
    return anim_slots[id].to;
}

/// Check if an animation has completed.
pub fn isAnimDone(id: usize) bool {
    return anim_slots[id].done;
}

/// Tick all active timing-based animations.
/// Call once per frame. `now_ms` = SDL_GetTicks() (milliseconds).
pub fn tickAnims(now_ms: u32) void {
    const now: f32 = @as(f32, @floatFromInt(now_ms)) / 1000.0;
    var any_active = false;

    for (0..anim_slot_count) |i| {
        var s = &anim_slots[i];
        if (!s.active) continue;

        // Set start time on first tick
        if (s.start_time < 0) {
            s.start_time = now;
        }

        const elapsed = (now - s.start_time) * 1000.0; // back to ms
        const duration = s.duration_ms;

        if (duration <= 0) {
            // Instant
            s.current = s.to;
            s.progress = 1.0;
            s.active = false;
            s.done = true;
            continue;
        }

        s.progress = std.math.clamp(elapsed / duration, 0.0, 1.0);
        const eased = s.easing(s.progress);
        s.current = lerp(s.from, s.to, eased);

        if (s.progress >= 1.0) {
            s.current = s.to;
            s.active = false;
            s.done = true;
        } else {
            any_active = true;
        }
    }

    _anims_active = any_active;
}

// ════════════════════════════════════════════════════════════════════════
// Phase 3: Spring Slots (physics-based)
// Reference: animate.lua:342-356 (state), animate.lua:468-493 (tick)
// ════════════════════════════════════════════════════════════════════════

pub const MAX_SPRING_SLOTS = 64;

pub const SpringSlot = struct {
    target: f32 = 0,
    current: f32 = 0,
    velocity: f32 = 0,
    stiffness: f32 = 100,
    damping: f32 = 10,
    mass: f32 = 1,
    rest_threshold: f32 = 0.01,
    active: bool = false,
};

var spring_slots: [MAX_SPRING_SLOTS]SpringSlot = [_]SpringSlot{.{}} ** MAX_SPRING_SLOTS;
var spring_slot_count: usize = 0;
var _springs_active: bool = false;

/// Create a spring-based animation slot. Returns the slot ID.
pub fn createSpring(stiffness: f32, damping: f32) usize {
    const id = spring_slot_count;
    std.debug.assert(id < MAX_SPRING_SLOTS);
    spring_slots[id] = .{
        .stiffness = stiffness,
        .damping = damping,
    };
    spring_slot_count += 1;
    return id;
}

/// Set the spring's target value. If different from current target,
/// the spring becomes active and starts moving toward it.
pub fn setSpringTarget(id: usize, target: f32) void {
    var s = &spring_slots[id];
    if (target != s.target or !s.active) {
        s.target = target;
        s.active = true;
        _springs_active = true;
    }
}

/// Set the spring's current value directly (for initialization).
pub fn setSpringValue(id: usize, value: f32) void {
    spring_slots[id].current = value;
}

/// Get the current spring value.
pub fn getSpringValue(id: usize) f32 {
    return spring_slots[id].current;
}

/// Check if a spring has reached rest.
pub fn isSpringAtRest(id: usize) bool {
    return !spring_slots[id].active;
}

/// Tick all active springs with Verlet integration.
/// `dt` is delta time in seconds (typically ~0.016 for 60fps).
/// Reference: animate.lua:468-493
pub fn tickSprings(dt_raw: f32) void {
    // Clamp dt to prevent physics explosion on frame hitches
    const dt = @min(dt_raw, 0.064);
    var any_active = false;

    for (0..spring_slot_count) |i| {
        var s = &spring_slots[i];
        if (!s.active) continue;

        const displacement = s.current - s.target;
        const spring_force = -s.stiffness * displacement;
        const damping_force = -s.damping * s.velocity;
        const acceleration = (spring_force + damping_force) / s.mass;

        s.velocity += acceleration * dt;
        s.current += s.velocity * dt;

        // Rest detection: snap to target when close enough and slow enough
        if (@abs(s.velocity) < s.rest_threshold and @abs(s.current - s.target) < s.rest_threshold) {
            s.current = s.target;
            s.velocity = 0;
            s.active = false;
        } else {
            any_active = true;
        }
    }

    _springs_active = any_active;
}

// ════════════════════════════════════════════════════════════════════════
// Global queries
// ════════════════════════════════════════════════════════════════════════

/// Returns true if any animation or spring is currently active.
/// Used by the main loop to decide whether to repaint even when
/// state hasn't changed.
pub fn isActive() bool {
    return _anims_active or _springs_active;
}

/// Reset all slots (for testing or hot-reload).
pub fn reset() void {
    anim_slot_count = 0;
    spring_slot_count = 0;
    _anims_active = false;
    _springs_active = false;
}
